Retries for network requests

Our tests on CI quite often fail because of TravisCI network instability. However, we shouldn't blame TravisCI, our clients could have the same fragile network. So, let's add retries for network requests.

Another requirement is to not introduce new dependencies for such a small task. Re-implementing our own ad-hoc [backoff](https://github.com/trendmicro/backoff-python) isn't hard.

```
tests/test_resolving/test_smoke.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dephell/controllers/_resolver.py:86: in resolve
    resolved = self._resolve(debug=debug, silent=silent, level=level, spinner=spinner)
dephell/controllers/_resolver.py:115: in _resolve
    no_conflicts = self._apply_deps(deps, debug=debug)
dephell/controllers/_resolver.py:195: in _apply_deps
    conflict = self.apply(dep)
dephell/controllers/_resolver.py:49: in apply
    if not other_dep.compat:
dephell/models/dependency.py:154: in compat
    for group in self.groups:
dephell/models/groups.py:180: in __iter__
    loop.run_until_complete(future)
/opt/pyenv/versions/3.8/lib/python3.8/asyncio/base_events.py:608: in run_until_complete
    return future.result()
dephell/models/groups.py:138: in _fetch_releases_deps
    await asyncio.gather(*tasks)
dephell/models/groups.py:96: in _fetch_missed_deps
    await asyncio.gather(
dephell/models/groups.py:96: in _fetch_missed_deps
    await asyncio.gather(
dephell/models/groups.py:84: in _fetch_missed_deps
    responses = await asyncio.gather(*tasks)
dephell/controllers/_repos.py:148: in get_dependencies
    return await repo.get_dependencies(name=name, version=version, extra=extra)
dephell/repositories/_warehouse/_api.py:122: in get_dependencies
    deps = await asyncio.gather(asyncio.ensure_future(task))
dephell/repositories/_warehouse/_api.py:259: in _get_from_json
    return await self._get_from_files(response['urls'])
dephell/repositories/_warehouse/_api.py:290: in _get_from_files
    return await self._download_and_parse(
dephell/repositories/_warehouse/_base.py:125: in _download_and_parse
    await self._download(url=url, path=path)
dephell/repositories/_warehouse/_base.py:148: in _download
    chunk = await response.content.read(1024)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <StreamReader 129565 bytes e=ClientPayloadError('Response payload is not completed')>
n = 1024
    async def read(self, n: int=-1) -> bytes:
        if self._exception is not None:
>           raise self._exception
E           aiohttp.client_exceptions.ClientPayloadError: Response payload is not completed
../../../.local/share/dephell/venvs/dephell-rRHZ/pytest/lib/python3.8/site-packages/aiohttp/streams.py:334: ClientPayloadError
----------------------------- Captured stdout call -----------------------------
```

--- a/dephell/cache.py
+++ b/dephell/cache.py
@@ -16,7 +16,7 @@ class BaseCache:
     def __init__(self, *keys, ttl: int = -1):
         self.path = Path(config['cache']['path'], *keys)
         if self.ext:
-            self.path = self.path.with_suffix(self.ext)
+            self.path = self.path.with_name(self.path.name + self.ext)
         self.ttl = ttl
         self._check_ttl()
 

--- a/dephell/commands/deps_tree.py
+++ b/dephell/commands/deps_tree.py
@@ -63,12 +63,18 @@ class DepsTreeCommand(BaseCommand):
 
     @classmethod
     def _make_tree(cls, dep, *, level: int = 0) -> List[str]:
-        lines = ['{level}- {name} [required: {constraint}, locked: {best}, latest: {latest}]'.format(
+        best = dep.group.best_release.version
+        latest = dep.group.best_release.version
+        if best == latest:
+            template = '{level}- {name} [required: {constraint}, latest: {latest}]'
+        else:
+            template = '{level}- {name} [required: {constraint}, locked: {best}, latest: {latest}]'
+        lines = [template.format(
             level='  ' * level,
             name=dep.name,
             constraint=str(dep.constraint) or '*',
-            best=str(dep.group.best_release.version),
-            latest=str(dep.groups.releases[0].version),
+            best=str(best),
+            latest=str(latest),
         )]
         deps = {dep.name: dep for dep in dep.dependencies}.values()  # drop duplicates
         for subdep in sorted(deps):

--- a/dephell/converters/egginfo.py
+++ b/dephell/converters/egginfo.py
@@ -2,6 +2,7 @@
 from collections import defaultdict
 from email.parser import Parser
 from itertools import chain
+from logging import getLogger
 from pathlib import Path
 from typing import Dict, Optional
 
@@ -9,7 +10,7 @@ from typing import Dict, Optional
 from dephell_discover import Root as PackageRoot
 from dephell_links import parse_link
 from dephell_markers import Markers
-from packaging.requirements import Requirement as PackagingRequirement
+from packaging.requirements import Requirement as PackagingRequirement, InvalidRequirement
 
 # app
 from ..constants import DOWNLOAD_FIELD, HOMEPAGE_FIELD
@@ -20,6 +21,7 @@ from .setuppy import SetupPyConverter
 
 
 class _Reader:
+    logger = getLogger('dephell.converters.egginfo')
 
     def can_parse(self, path: Path, content: Optional[str] = None) -> bool:
         if isinstance(path, str):
@@ -151,7 +153,15 @@ class _Reader:
         # dependencies
         deps = []
         for req in cls._get_list(info, 'Requires-Dist'):
-            req = PackagingRequirement(req)
+            try:
+                req = PackagingRequirement(req)
+            except InvalidRequirement:
+                cls.logger.warning('invalid requirement', extra=dict(
+                    requirement=req,
+                    package_name=root.name,
+                    package_version=root.version,
+                ))
+                continue
             deps.extend(DependencyMaker.from_requirement(
                 source=root,
                 req=req,

--- a/dephell/converters/egginfo.py
+++ b/dephell/converters/egginfo.py
@@ -10,7 +10,7 @@ from typing import Dict, Optional
 from dephell_discover import Root as PackageRoot
 from dephell_links import parse_link
 from dephell_markers import Markers
-from packaging.requirements import Requirement as PackagingRequirement, InvalidRequirement
+from packaging.requirements import InvalidRequirement, Requirement as PackagingRequirement
 
 # app
 from ..constants import DOWNLOAD_FIELD, HOMEPAGE_FIELD

--- a/dephell/models/dependency.py
+++ b/dephell/models/dependency.py
@@ -193,7 +193,7 @@ class Dependency:
 
         marker = Markers(str(self.marker))
         if self.envs - {'main'}:
-            extra_markers = {'extra == "{}"'.format(env) for env in self.envs - {'main'}}
+            extra_markers = {'extra == {!r}'.format(env) for env in self.envs - {'main'}}
             marker &= Markers(' or '.join(extra_markers))
         if marker:
             result += '; ' + str(marker)

--- a/dephell/networking.py
+++ b/dephell/networking.py
@@ -1,16 +1,20 @@
 # built-in
+from functools import partial, update_wrapper
+from logging import getLogger
+from time import sleep
 from ssl import create_default_context
 
 # external
 import certifi
 import requests
-from aiohttp import ClientSession, TCPConnector
+from aiohttp import ClientSession, TCPConnector, ClientPayloadError
 
 # app
 from . import __version__
 
 
 USER_AGENT = 'DepHell/{version}'.format(version=__version__)
+logger = getLogger('dephell.networking')
 
 
 def aiohttp_session(*, auth=None, **kwargs):
@@ -36,3 +40,22 @@ def requests_session(*, auth=None, headers=None, **kwargs):
     if kwargs:
         session.__dict__.update(kwargs)
     return session
+
+
+def aiohttp_repeat(func=None, *, count: int = 3):
+    if func is None:
+        return partial(func, count=count)
+
+    async def wrapper(*args, **kwargs):
+        for pause in range(1, count + 1):
+            try:
+                return await func(*args, **kwargs)
+            except ClientPayloadError:
+                if pause == count:
+                    raise
+                logger.debug('aiohttp payload error, repeating...', exc_info=True)
+                sleep(pause)
+        raise RuntimeError('unreachable')
+
+    wrapper = update_wrapper(wrapper=wrapper, wrapped=func)
+    return wrapper

--- a/dephell/networking.py
+++ b/dephell/networking.py
@@ -7,7 +7,7 @@ from ssl import create_default_context
 # external
 import certifi
 import requests
-from aiohttp import ClientSession, TCPConnector, ClientPayloadError
+from aiohttp import ClientSession, TCPConnector, ClientError
 
 # app
 from . import __version__
@@ -50,7 +50,7 @@ def aiohttp_repeat(func=None, *, count: int = 4):
         for pause in range(1, count + 1):
             try:
                 return await func(*args, **kwargs)
-            except ClientPayloadError:
+            except ClientError:
                 if pause == count:
                     raise
                 logger.debug('aiohttp payload error, repeating...', exc_info=True)

--- a/dephell/networking.py
+++ b/dephell/networking.py
@@ -1,13 +1,13 @@
 # built-in
 from functools import partial, update_wrapper
 from logging import getLogger
-from time import sleep
 from ssl import create_default_context
+from time import sleep
 
 # external
 import certifi
 import requests
-from aiohttp import ClientSession, TCPConnector, ClientError
+from aiohttp import ClientError, ClientSession, TCPConnector
 
 # app
 from . import __version__

--- a/dephell/networking.py
+++ b/dephell/networking.py
@@ -42,7 +42,7 @@ def requests_session(*, auth=None, headers=None, **kwargs):
     return session
 
 
-def aiohttp_repeat(func=None, *, count: int = 3):
+def aiohttp_repeat(func=None, *, count: int = 4):
     if func is None:
         return partial(func, count=count)
 

--- a/dephell/repositories/_warehouse/_base.py
+++ b/dephell/repositories/_warehouse/_base.py
@@ -99,22 +99,23 @@ class WarehouseBaseRepo(Interface):
                     ))
 
             try:
-                dep_extra = req.marker and Markers(req.marker).extra
+                dep_extras = req.marker and Markers(req.marker).get_strings('extra')
             except ValueError:  # unsupported operation for version marker python_version: in
-                dep_extra = None
+                dep_extras = set()
 
             # it's not extra and we want not extra too
-            if dep_extra is None and extra is None:
+            if not dep_extras and extra is None:
                 result.append(req)
                 continue
             # it's extra, but we want not the extra
             # or it's not the extra, but we want extra.
-            if dep_extra is None or extra is None:
+            if not dep_extras or extra is None:
                 continue
             # it's extra and we want this extra
-            elif dep_extra == extra:
-                result.append(req)
-                continue
+            for dep_extra in dep_extras:
+                if dep_extra == extra:
+                    result.append(req)
+                    break
 
         return tuple(result)
 

--- a/dephell/repositories/_warehouse/_base.py
+++ b/dephell/repositories/_warehouse/_base.py
@@ -13,7 +13,7 @@ from packaging.requirements import InvalidRequirement, Requirement
 # app
 from ...cached_property import cached_property
 from ...constants import WAREHOUSE_DOMAINS
-from ...networking import aiohttp_session, aiohttp_repeat
+from ...networking import aiohttp_repeat, aiohttp_session
 from ..base import Interface
 
 

--- a/dephell/repositories/_warehouse/_base.py
+++ b/dephell/repositories/_warehouse/_base.py
@@ -13,7 +13,7 @@ from packaging.requirements import InvalidRequirement, Requirement
 # app
 from ...cached_property import cached_property
 from ...constants import WAREHOUSE_DOMAINS
-from ...networking import aiohttp_session
+from ...networking import aiohttp_session, aiohttp_repeat
 from ..base import Interface
 
 
@@ -136,6 +136,7 @@ class WarehouseBaseRepo(Interface):
                         deps.append(str(dep))
             return tuple(deps)
 
+    @aiohttp_repeat
     async def _download(self, *, url: str, path: Path) -> None:
         async with aiohttp_session(auth=self.auth) as session:
             async with session.get(url) as response:

--- a/dephell/repositories/_warehouse/_simple.py
+++ b/dephell/repositories/_warehouse/_simple.py
@@ -143,7 +143,8 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
         )
         links = cache.load()
         if links:
-            return links
+            yield from links
+            return
 
         dep_url = posixpath.join(self.url, quote(name)) + '/'
         with requests_session() as session:

--- a/dephell/repositories/_warehouse/_simple.py
+++ b/dephell/repositories/_warehouse/_simple.py
@@ -211,9 +211,6 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
             (sdist, '.zip'),
         )
 
-        if name == 'celery':
-            import pdb; pdb.set_trace()
-
         for converter, ext in rules:
             for link in good_links:
                 if not link['name'].endswith(ext):

--- a/dephell/repositories/_warehouse/_simple.py
+++ b/dephell/repositories/_warehouse/_simple.py
@@ -53,9 +53,19 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
         releases_info = dict()
         for link in links:
             name, version = self._parse_name(link['name'])
-            if canonicalize_name(name) != dep.name:
+            if canonicalize_name(name) != canonicalize_name(dep.base_name):
+                logger.warning('bad dist name', extra=dict(
+                    dist_name=link['name'],
+                    package_name=dep.base_name,
+                    reason='package name does not match',
+                ))
                 continue
             if not version:
+                logger.warning('bad dist name', extra=dict(
+                    dist_name=link['name'],
+                    package_name=dep.base_name,
+                    reason='no version specified',
+                ))
                 continue
 
             if version not in releases_info:

--- a/dephell/repositories/_warehouse/_simple.py
+++ b/dephell/repositories/_warehouse/_simple.py
@@ -147,6 +147,7 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
 
         dep_url = posixpath.join(self.url, quote(name)) + '/'
         with requests_session() as session:
+            logger.debug('getting dep info from simple repo', extra=dict(url=dep_url))
             response = session.get(dep_url, auth=self.auth)
         if response.status_code == 404:
             raise PackageNotFoundError(package=name, url=dep_url)
@@ -164,12 +165,14 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
 
             python = tag.get('data-requires-python')
             fragment = parse_qs(parsed.fragment)
-            yield dict(
+            link = dict(
                 url=urljoin(dep_url, link),
                 name=parsed.path.strip('/').split('/')[-1],
                 python=html.unescape(python) if python else '*',
                 digest=fragment['sha256'][0] if 'sha256' in fragment else None,
             )
+            links.append(link)
+            yield link
 
         cache.dump(links)
         return links

--- a/dephell/repositories/_warehouse/_simple.py
+++ b/dephell/repositories/_warehouse/_simple.py
@@ -152,7 +152,7 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
             ttl=config['cache']['ttl'],
         )
         links = cache.load()
-        if links:
+        if links is not None:
             yield from links
             return
 

--- a/dephell/repositories/_warehouse/_simple.py
+++ b/dephell/repositories/_warehouse/_simple.py
@@ -188,7 +188,7 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
         cache.dump(links)
         return links
 
-    async def _get_deps_from_links(self, name, version):
+    async def _get_deps_from_links(self, name: str, version):
         from ...converters import SDistConverter, WheelConverter
 
         links = self._get_links(name=name)
@@ -197,7 +197,7 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
             link_name, link_version = self._parse_name(link['name'])
             if canonicalize_name(link_name) != name:
                 continue
-            if link_version != version:
+            if link_version != str(version):
                 continue
             good_links.append(link)
 
@@ -210,6 +210,9 @@ class WarehouseSimpleRepo(WarehouseBaseRepo):
             (sdist, '.tar.gz'),
             (sdist, '.zip'),
         )
+
+        if name == 'celery':
+            import pdb; pdb.set_trace()
 
         for converter, ext in rules:
             for link in good_links:

--- a/docs/conf.py
+++ b/docs/conf.py
@@ -7,7 +7,6 @@ from pathlib import Path
 
 # external
 import alabaster
-from recommonmark.parser import CommonMarkParser
 from recommonmark.transform import AutoStructify
 
 
@@ -20,12 +19,10 @@ extensions = [
     'sphinx.ext.coverage',
     'sphinx.ext.viewcode',
     'sphinx.ext.githubpages',
+    'recommonmark',
 ]
 
 templates_path = ['_templates']
-source_parsers = {
-    '.md': CommonMarkParser,
-}
 source_suffix = ['.rst', '.md']
 master_doc = 'index'
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -11,3 +11,8 @@ exclude=
     .pytest_cache
     build
     setup.py
+
+[tool:pytest]
+addopts = --strict-markers
+markers =
+    allow_hosts: allow network requests to specified hostnames.

--- a/tests/test_controllers/test_safety.py
+++ b/tests/test_controllers/test_safety.py
@@ -9,5 +9,5 @@ from dephell.controllers import Safety
 def test_safety():
     safety = Safety()
     vulns = safety.get('django', '1.11.0')
-    assert len(vulns) == 5
+    assert len(vulns) == 13
     assert {vuln.name for vuln in vulns} == {'django'}

